jdbc-driver/src/main/java/com/google/cloud/dataproc/jdbc/DataprocDriverUtils.java

```diff
@@ -0,0 +1,246 @@
+                hiveJdbcURL = String.format("%s%s", hiveJdbcURL, otherSessionConfs());
+            if (hiveConfs() != null) hiveJdbcURL = String.format("%s?%s", hiveJdbcURL, hiveConfs());
+            if (hiveVars() != null) hiveJdbcURL = String.format("%s#%s", hiveJdbcURL, hiveVars());
+            return hiveJdbcURL;
+        }
+
+        @AutoValue.Builder
+        abstract static class Builder {
+            abstract Builder setHost(String value);
+
+            abstract Builder setPort(int value);
+
+            abstract Builder setDbName(String value);
+
+            abstract Builder setHttpPath(String value);
+
+            abstract Builder setTransportMode(String value);
+
+            abstract Builder setOtherSessionConfs(String value);
+
+            abstract Builder setHiveConfs(String value);
+
+            abstract Builder setHiveVars(String value);
+
+            abstract HiveJdbcConnectionParam build();
+        }
+    }
+
+    /**
+     * Parses client url and extracts Hive connection parameters.
+     *
+     * @param url possible formats:
+     *     jdbc:dataproc://hive/<dbName>;clusterName=<>;projectId=<>;region=<>;port=<>
+     *     jdbc:dataproc://hive/<dbName>;clusterName=<>;projectId=<>;region=<>;clusterPoolLabel=foo=bar:env=staging;
+     * @return HiveJdbcConnectionParam object with fields correctly set
+     * @throws IOException
+     * @throws InvalidURLException
+     */
+    public static HiveJdbcConnectionParam parseHiveUrl(String url) throws InvalidURLException {
+        HiveJdbcConnectionParam.Builder paramBuilder = HiveJdbcConnectionParam.builder();
+        String urlStripped = url.substring(URL_HIVE_PREFIX.length());
+
+        String dbName = urlStripped.substring(0, urlStripped.indexOf(';'));
+        paramBuilder.setDbName(dbName.isEmpty() ? DEFAULT_DATABASE : dbName);
+
+        int hiveConfsStart = urlStripped.indexOf('?');
+        int hiveVarsStart = urlStripped.indexOf("#");
+
+        String hiveConfs =
+                hiveConfsStart == -1
+                        ? null
+                        : urlStripped.substring(
+                        hiveConfsStart + 1, hiveVarsStart == -1 ? urlStripped.length() : hiveVarsStart);
+        String hiveVars = hiveVarsStart == -1 ? null : urlStripped.substring(hiveVarsStart + 1);
+
+        // Params that do not need parsing
+        paramBuilder.setHiveConfs(hiveConfs);
+        paramBuilder.setHiveVars(hiveVars);
+
+        // Parse sessionConfigs to get cluster info
+        String sessionConfigs = urlStripped;
+        if (hiveConfsStart != -1 || hiveVarsStart != -1) {
+            sessionConfigs =
+                    urlStripped.substring(0, hiveConfsStart != -1 ? hiveConfsStart : hiveVarsStart);
+        }
+
+        // key=value pattern
+        Pattern pattern = Pattern.compile("([^;]*)=([^;]*)[;]?");
+        Matcher matcher = pattern.matcher(sessionConfigs);
+        StringBuilder sb = new StringBuilder();
```

Comment:
Nit: rename `sb` to a clearer variable name. (It might disappear if you end up changing this parsing logic, though)